components/Banner/examples/supplemental/MDCBannerView.m

```diff
@@ -54,124 +67,527 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
-        MDCButton *button = [[MDCButton alloc] init];
-        [mutableButtons addObject:button];
-        [self addSubview:button];
-      }
-    } else if (numberOfButtons < self.buttons.count) {
-      for (NSUInteger index = numberOfButtons; index < self.buttons.count; ++index) {
-        MDCButton *lastObject = [mutableButtons lastObject];
-        [lastObject removeFromSuperview];
-        [mutableButtons removeLastObject];
-      }
-    }
-    self.buttons = [mutableButtons copy];
+- (UIImage *)icon {
+  return self.iconImageView.image;
+}
+
+- (void)setButtons:(NSMutableArray<MDCButton *> *_Nonnull)buttons {
+  NSAssert(buttons.count <= kNumberOfButtonsLimit, @"%@ doesn't support more than %lu buttons",
+           NSStringFromClass([self class]), (unsigned long)kNumberOfButtonsLimit);
+  _buttons = buttons;
+  for (MDCButton *button in buttons) {
+    button.translatesAutoresizingMaskIntoConstraints = NO;
+    [self.containerView addSubview:button];
   }
+  self.isLayoutDirty = YES;
+}
+
+- (void)setBackgroundColor:(UIColor *)backgroundColor {
+  [super setBackgroundColor:backgroundColor];
+  self.containerView.backgroundColor = backgroundColor;
+}
+
+- (void)setTextFont:(UIFont *)textFont {
+  self.textLabel.font = textFont;
+}
+
+- (UIFont *)textFont {
+  return self.textLabel.font;
+}
+
+- (void)setTextColor:(UIColor *)textColor {
+  self.textLabel.textColor = textColor;
+}
+
+- (UIColor *)textColor {
+  return self.textLabel.textColor;
 }
 
-- (NSUInteger)numberOfButtons {
-  return self.buttons.count;
+- (void)setIconTintColor:(UIColor *)iconTintColor {
+  self.iconImageView.tintColor = iconTintColor;
+}
+
+- (UIColor *)iconTintColor {
+  return self.iconImageView.tintColor;
 }
 
 #pragma mark - UIView overrides
 
 - (CGSize)sizeThatFits:(CGSize)size {
-  self.layout = [[MDCBannerViewLayout alloc] initWithSizeToFit:size
-                                                     textLabel:self.textLabel
-                                                imageContainer:self.iconImageViewContainer
-                                                       buttons:self.buttons];
-  return [self.layout frameSize];
+  return CGSizeMake([UIScreen mainScreen].bounds.size.width, self.layout.size.height);
 }
 
-- (void)layoutSubviews {
-  [super layoutSubviews];
+- (CGSize)intrinsicContentSize {
+  return CGSizeMake([UIScreen mainScreen].bounds.size.width, self.layout.size.height);
+}
 
-  [self applyMDCBannerViewLayout:self.layout];
+- (void)updateConstraints {
+  if (!self.cachedLayout || self.cachedLayout != self.layout) {
+    [self updateConstraintsWithLayoutStyle:self.layout.style];
+    self.cachedLayout = self.layout;
+  }
+
+  [super updateConstraints];
 }
 
 #pragma mark - Layout methods
 
-- (void)applyMDCBannerViewLayout:(MDCBannerViewLayout *)layout {
-  if (![self isMDCBannerViewLayoutApplicable:layout]) {
-    return;
+- (void)setIsLayoutDirty:(BOOL)isLayoutDirty {
```

Comment:
I believe we can remove this and replace where needed with `[self invalidateIntrinsicContentSize]`

Reply:
`invalidateIntrinsicContentSize` will make sure `intrinsicContentSize` is called, but I need the `layout` also get recalculated before that. This flag ensures `layout` will be lazily recalculated in this case. I think there is some area to make the code cleaner and more readable. Any suggestion would be welcome and we can talk offline.

Reply:
Please make sure `intrinsicContentSize` does NOT depend on the current layout. It can depend on the currently set properties, but not the actual layout. Similar for `-sizeThatFits:` - it should not depend on the current position/size of views.

Reply:
@romoore Thanks for your comment. It is not dependent of the current layout. I think my `layout` property might be confusing by its word, but it is calculated separately.

Reply:
I became worried from your statement that, "I need the layout also get recalculated before [`intrinsicContentSize` is called]."  I would expect that `intrinsicContentSize to be invariant of any layout work and would call any necessary methods before calculating size.

Reply:
The `instrinsicContentSize` is relevant to some `layout` work. The word I used might be not appropriate. For example,  `instrinsicContentSize` depends on how many buttons users set in the view and whether these buttons can fit into a user specified width. Does it sound reasonable? Any suggestion would be good.

Reply:
It's OK to have overlap between `layoutSubviews` and `intrinsicContentSize`. In some cases the code can be in a reused method, other times it's simply duplicated.   For example, [in a view I wrote for a snapshot test, the `layoutSubviews` and `sizeThatFits:` code is nearly identical](https://github.com/material-components/material-components-ios/pull/6828/files#diff-6415a9a7f22a15e5ce8f8d9a400ba6adR84).  In [MDCBottomNavigationBar both the layout code and `intrinsicContentSize` use a shared method to compute the expected width](https://github.com/material-components/material-components-ios/blob/develop/components/BottomNavigation/src/MDCBottomNavigationBar.m#L191).

I'm not concerned that the calculated size depends on the state of the view&mdash;that's normal. I'm concerned that something called `layout` needs to be recalculated before you can determine the size.  It seems like that `layout` object is the current layout on the screen, and  that the on-screen layout would have nothing at all to do with the potential size returned by either `sizeThatFits:` or `intrinsicContentSize`.

Reply:
Thanks for the examples. I don't think my code here conflicts with the concept you mentioned, and it aligns with your examples. The `layout` is probably a bad name. What I did was basically extracting the `widthForItemsWhenCenteredWithAvailableWidth:CGFLOAT_MAX` method in your [example](https://github.com/material-components/material-components-ios/blob/develop/components/BottomNavigation/src/MDCBottomNavigationBar.m#L193) into a separate object called `layout`. Recalculating it is basically calling this method though the logic is a little more complex, which was why it was extracted. I should probably change the name or moving this logic into the same class before/after design review. What do you think?

Reply:
At this point, I think you should split your refactoring changes (like creating a `layout` object) away from the improvements you're making for RTL support.  There are several long threads on this PR and I think that's another good indication that it's probably too large.

Reply:
I agree with it is a little large on the scale. My refactor was adding Autolayout with corresponding refactors. The ability to support RTL is a result of it. So they can kinda coupled. Would you be okay if this PR land as a whole?

Reply:
I don't see a specific reason why these need to land together and the more "conversational" changes have also held-up the more routine refactorings/improvements. I still believe this should be split into multiple PRs.

Seems like there are at least 3 changes here:
* Autolayout (support?)
* RTL support
* One or more internal clean-up/refactoring changes